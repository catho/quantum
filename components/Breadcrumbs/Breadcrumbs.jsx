import { useState, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import { theme as defaultTheme } from '../shared';
import Icon from '../Icon';
import Link from '../Link';

const BreadcrumbsItems = styled.ol`
  display: flex;
  align-items: center;

  list-style: none;

  overflow: auto;

  ${({
    theme: {
      spacing: { xxsmall, small },
    },
  }) => css`
    padding: ${small}px ${xxsmall}px;
  `}
`;

const BreadcrumbItem = styled.li`
  display: flex;
  align-items: center;

  ${({
    theme: {
      baseFontSize,
      colors: { neutral },
      spacing: { xsmall },
    },
  }) => css`
    font-size: ${baseFontSize * 0.75}px;

    a {
      display: flex;
      align-items: center;

      font-weight: 400;

      &:hover {
        text-decoration: underline;
      }
    }

    span {
      color: ${neutral[700]};
    }

    & + li {
      margin-left: ${xsmall}px;
    }

    &[aria-hidden='true'] {
      border: 0;
      height: 0;
      margin: 0;
      opacity: 0;
      overflow: hidden;
      padding: 0;
      position: absolute;
      width: 0;
      visibility: hidden;
    }
  `}
`;

const BreadcrumbIcon = styled(Icon).attrs(
  ({ name = 'keyboard_arrow_right' }) => ({
    name,
    size: 'small',
    'aria-hidden': true,
  }),
)`
  ${({
    theme: {
      colors: { primary },
      spacing: { xxsmall, xsmall },
    },
    name,
  }) => css`
    margin-right: ${name === 'home' ? xxsmall : xsmall}px;
    color: ${primary[700]};
  `}
`;

const ExpandButton = styled.button.attrs(() => ({
  type: 'button',
}))`
  border: none;
  background-color: transparent;

  cursor: pointer;

  &:hover {
    font-weight: 700;
  }

  ${({
    theme: {
      colors: { primary },
    },
  }) => css`
    color: ${primary[700]};
  `}
`;

const Breadcrumbs = ({ items, theme }) => {
  if (items.length < 2) {
    // eslint-disable-next-line no-console
    console.error('Breadcrumbs items should be greater or equal than 2.');

    return null;
  }

  const [collapsed, setCollapsed] = useState(false);
  const [secondLink, setSecondLink] = useState(null);

  const breadcrumbsItemsEl = useRef(null);

  const lastItemIndex = items.length - 1;

  const [firstItem, ...restItems] = items;
  const lastItem = items[lastItemIndex];

  restItems.splice(lastItemIndex - 1, 1);

  useEffect(() => {
    if (items.length > 2 && breadcrumbsItemsEl.current) {
      const { scrollWidth, clientWidth } = breadcrumbsItemsEl.current;

      setCollapsed(clientWidth < scrollWidth);
    }
  }, []);

  useEffect(() => {
    if (secondLink) {
      secondLink.focus();
    }
  }, [secondLink]);

  const handleExpand = () => {
    setCollapsed(false);

    const focusableItem = breadcrumbsItemsEl.current.querySelector(
      'li[aria-hidden="true"] a',
    );

    setSecondLink(focusableItem);
  };

  return (
    <nav>
      <BreadcrumbsItems
        id="breadcrumbs-list"
        ref={breadcrumbsItemsEl}
        aria-label="breadcrumb"
        theme={theme}
      >
        <BreadcrumbItem theme={theme}>
          <Link underline={false} href={firstItem.url} title={firstItem.label}>
            <BreadcrumbIcon theme={theme} name="home" />
            {firstItem.label}
          </Link>
        </BreadcrumbItem>

        <BreadcrumbItem
          theme={theme}
          hidden={!collapsed}
          aria-hidden={!collapsed}
        >
          <BreadcrumbIcon theme={theme} />
          <ExpandButton
            theme={theme}
            aria-expanded={!collapsed}
            aria-controls="breadcrumbs-list"
            aria-label="Exibir todas as opções"
            title="Exibir todas as opções"
            onClick={handleExpand}
          >
            ...
          </ExpandButton>
        </BreadcrumbItem>

        {restItems.map((item) => (
          <BreadcrumbItem
            key={item.label}
            theme={theme}
            aria-hidden={collapsed}
          >
            <BreadcrumbIcon theme={theme} />
            <Link underline={false} href={item.url} title={item.label}>
              {item.label}
            </Link>
          </BreadcrumbItem>
        ))}

        <BreadcrumbItem theme={theme}>
          <BreadcrumbIcon theme={theme} />
          <span>{lastItem.label}</span>
        </BreadcrumbItem>
      </BreadcrumbsItems>
    </nav>
  );
};

Breadcrumbs.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      url: PropTypes.string,
    }),
  ).isRequired,
  theme: PropTypes.shape({
    colors: PropTypes.object,
    spacing: PropTypes.object,
    baseFontSize: PropTypes.number,
  }),
};

Breadcrumbs.defaultProps = {
  theme: defaultTheme,
};

export default Breadcrumbs;
