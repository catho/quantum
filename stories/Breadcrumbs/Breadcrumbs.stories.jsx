// Generated with scripts/create-componnent.js

import Breadcrumbs from '../../components/Breadcrumbs';
import defaultContent from '../../test-utils/mocks/Breadcrumb/exampleContents';

export default {
  title: 'Breadcrumbs',
  component: Breadcrumbs,
};

const Template = (args) => <Breadcrumbs items={defaultContent} {...args} />;

export const Default = Template.bind({});
