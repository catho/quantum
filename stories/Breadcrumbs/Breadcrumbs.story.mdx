import { Meta, Canvas, Story, ArgsTable } from '@storybook/addon-docs';

import Breadcrumbs from '../../components/Breadcrumbs';

import { Header } from '../shared';

import { 
  Default
} from './Breadcrumbs.stories.jsx';

<Meta title="Breadcrumbs" component={Breadcrumbs} />

<Header title="Breadcrumbs">
  Breadcrumbs component consists of a list of links that helps an user to visually find themselves and the current page within the hierarchical structure of a website. It also allows the navigation up to any of its "ancestors".
</Header>

- [Importing and usage](#importing)
- [API](#api)
- [Example](#example)

## <a id="importing"></a>Importing and usage

```js
import Breadcrumbs from '@catho/quantum/Breadcrumbs';
```



## <a id="api"></a>API

<ArgsTable of={Breadcrumbs} />


## <a id="example"></a>Example

<Canvas>
  <Story story={Default} />
</Canvas>
