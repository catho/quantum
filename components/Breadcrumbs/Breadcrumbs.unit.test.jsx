import { render, fireEvent, screen } from '@testing-library/react';

import Breadcrumbs from './Breadcrumbs';
import defaultContent from '../../test-utils/mocks/Breadcrumb/exampleContents';

const mockScrollWidth = (value = 450) => {
  // eslint-disable-next-line no-undef
  Object.defineProperty(HTMLElement.prototype, 'scrollWidth', {
    configurable: true,
    value,
  });
};

const mockWindowInnerWidth = (value = 360) => {
  Object.assign(window, { innerWidth: value });
};

describe('<Breadcrumbs />', () => {
  it('should match the snapshots', () => {
    const { asFragment } = render(<Breadcrumbs items={defaultContent} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should match the snapshots when breadcrumbs is collapsed', () => {
    mockScrollWidth();
    mockWindowInnerWidth();

    const { asFragment } = render(<Breadcrumbs items={defaultContent} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should match the snapshots when collapsed breadcrumbs is expanded', () => {
    mockScrollWidth();
    mockWindowInnerWidth();

    const { getByText, asFragment } = render(
      <Breadcrumbs items={defaultContent} />,
    );

    const firstRender = asFragment();

    fireEvent.click(getByText('...'));

    expect(firstRender).toMatchDiffSnapshot(asFragment());
  });

  it('should all links have href attribute', () => {
    render(<Breadcrumbs items={defaultContent} />);

    const links = screen.getAllByRole('link');

    const hasLinkWithoutHREFAttribute = links.some(
      (link) => !link.hasAttribute('href'),
    );

    expect(hasLinkWithoutHREFAttribute).toBe(false);
  });

  it('should receive an error when items length less than 2', () => {
    const consoleError = jest.spyOn(console, 'error').mockImplementation();

    const { container } = render(
      <Breadcrumbs items={[{ label: 'some-label', url: 'some-url' }]} />,
    );

    expect(container.firstChild).toBeNull();
    expect(consoleError).toHaveBeenCalled();

    jest.restoreAllMocks();
  });
});
