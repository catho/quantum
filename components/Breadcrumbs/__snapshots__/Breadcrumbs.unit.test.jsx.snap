// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumbs /> should match the snapshots 1`] = `
<DocumentFragment>
  .c2 {
  color: #1250C4;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  font-weight: 700;
  word-break: break-all;
}

.c2:hover {
  color: #DE0059;
}

.c2:visited {
  color: #5734BA;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  list-style: none;
  overflow: auto;
  padding: 12px 4px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 12px;
}

.c1 a {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 400;
}

.c1 a:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1 span {
  color: #424242;
}

.c1 + li {
  margin-left: 8px;
}

.c1[aria-hidden='true'] {
  border: 0;
  height: 0;
  margin: 0;
  opacity: 0;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 0;
  visibility: hidden;
}

.c3 {
  margin-right: 4px;
  color: #1250C4;
}

.c4 {
  margin-right: 8px;
  color: #1250C4;
}

.c5 {
  border: none;
  background-color: transparent;
  cursor: pointer;
  color: #1250C4;
}

.c5:hover {
  font-weight: 700;
}

<nav>
    <ol
      aria-label="breadcrumb"
      class="c0"
      id="breadcrumbs-list"
    >
      <li
        class="c1"
      >
        <a
          class="c2"
          href="catho.com"
          title="Home"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root c3"
            data-qtm-preloader="icon"
            focusable="false"
            style="font-size: 16px; max-width: 16px;"
            theme="[object Object]"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
          Home
        </a>
      </li>
      <li
        aria-hidden="true"
        class="c1"
        hidden=""
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root c4"
          data-qtm-preloader="icon"
          focusable="false"
          style="font-size: 16px; max-width: 16px;"
          theme="[object Object]"
          viewBox="0 0 24 24"
        >
          <path
            d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
          />
        </svg>
        <button
          aria-controls="breadcrumbs-list"
          aria-expanded="true"
          aria-label="Exibir todas as opções"
          class="c5"
          title="Exibir todas as opções"
          type="button"
        >
          ...
        </button>
      </li>
      <li
        aria-hidden="false"
        class="c1"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root c4"
          data-qtm-preloader="icon"
          focusable="false"
          style="font-size: 16px; max-width: 16px;"
          theme="[object Object]"
          viewBox="0 0 24 24"
        >
          <path
            d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
          />
        </svg>
        <a
          class="c2"
          href="catho.com"
          title="Quantum"
        >
          Quantum
        </a>
      </li>
      <li
        aria-hidden="false"
        class="c1"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root c4"
          data-qtm-preloader="icon"
          focusable="false"
          style="font-size: 16px; max-width: 16px;"
          theme="[object Object]"
          viewBox="0 0 24 24"
        >
          <path
            d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
          />
        </svg>
        <a
          class="c2"
          href="catho.com"
          title="Components"
        >
          Components
        </a>
      </li>
      <li
        class="c1"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root c4"
          data-qtm-preloader="icon"
          focusable="false"
          style="font-size: 16px; max-width: 16px;"
          theme="[object Object]"
          viewBox="0 0 24 24"
        >
          <path
            d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
          />
        </svg>
        <span>
          Breadcrumbs
        </span>
      </li>
    </ol>
  </nav>
</DocumentFragment>
`;

exports[`<Breadcrumbs /> should match the snapshots when breadcrumbs is collapsed 1`] = `
<DocumentFragment>
  .c2 {
  color: #1250C4;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  font-weight: 700;
  word-break: break-all;
}

.c2:hover {
  color: #DE0059;
}

.c2:visited {
  color: #5734BA;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  list-style: none;
  overflow: auto;
  padding: 12px 4px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 12px;
}

.c1 a {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 400;
}

.c1 a:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1 span {
  color: #424242;
}

.c1 + li {
  margin-left: 8px;
}

.c1[aria-hidden='true'] {
  border: 0;
  height: 0;
  margin: 0;
  opacity: 0;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 0;
  visibility: hidden;
}

.c3 {
  margin-right: 4px;
  color: #1250C4;
}

.c4 {
  margin-right: 8px;
  color: #1250C4;
}

.c5 {
  border: none;
  background-color: transparent;
  cursor: pointer;
  color: #1250C4;
}

.c5:hover {
  font-weight: 700;
}

<nav>
    <ol
      aria-label="breadcrumb"
      class="c0"
      id="breadcrumbs-list"
    >
      <li
        class="c1"
      >
        <a
          class="c2"
          href="catho.com"
          title="Home"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root c3"
            data-qtm-preloader="icon"
            focusable="false"
            style="font-size: 16px; max-width: 16px;"
            theme="[object Object]"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
          Home
        </a>
      </li>
      <li
        aria-hidden="false"
        class="c1"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root c4"
          data-qtm-preloader="icon"
          focusable="false"
          style="font-size: 16px; max-width: 16px;"
          theme="[object Object]"
          viewBox="0 0 24 24"
        >
          <path
            d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
          />
        </svg>
        <button
          aria-controls="breadcrumbs-list"
          aria-expanded="false"
          aria-label="Exibir todas as opções"
          class="c5"
          title="Exibir todas as opções"
          type="button"
        >
          ...
        </button>
      </li>
      <li
        aria-hidden="true"
        class="c1"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root c4"
          data-qtm-preloader="icon"
          focusable="false"
          style="font-size: 16px; max-width: 16px;"
          theme="[object Object]"
          viewBox="0 0 24 24"
        >
          <path
            d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
          />
        </svg>
        <a
          class="c2"
          href="catho.com"
          title="Quantum"
        >
          Quantum
        </a>
      </li>
      <li
        aria-hidden="true"
        class="c1"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root c4"
          data-qtm-preloader="icon"
          focusable="false"
          style="font-size: 16px; max-width: 16px;"
          theme="[object Object]"
          viewBox="0 0 24 24"
        >
          <path
            d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
          />
        </svg>
        <a
          class="c2"
          href="catho.com"
          title="Components"
        >
          Components
        </a>
      </li>
      <li
        class="c1"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root c4"
          data-qtm-preloader="icon"
          focusable="false"
          style="font-size: 16px; max-width: 16px;"
          theme="[object Object]"
          viewBox="0 0 24 24"
        >
          <path
            d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
          />
        </svg>
        <span>
          Breadcrumbs
        </span>
      </li>
    </ol>
  </nav>
</DocumentFragment>
`;

exports[`<Breadcrumbs /> should match the snapshots when collapsed breadcrumbs is expanded 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -28,12 +28,13 @@
            </svg>
            Home
          </a>
        </li>
        <li
-         aria-hidden=\\"false\\"
+         aria-hidden=\\"true\\"
          class=\\"Breadcrumbs__BreadcrumbItem-sc-d5msik-1 cWkJmI\\"
+         hidden=\\"\\"
        >
          <svg
            aria-hidden=\\"true\\"
            class=\\"MuiSvgIcon-root Breadcrumbs__BreadcrumbIcon-sc-d5msik-2 jEpugD\\"
            data-qtm-preloader=\\"icon\\"
@@ -46,21 +47,21 @@
              d=\\"M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z\\"
            />
          </svg>
          <button
            aria-controls=\\"breadcrumbs-list\\"
-           aria-expanded=\\"false\\"
+           aria-expanded=\\"true\\"
            aria-label=\\"Exibir todas as opções\\"
            class=\\"Breadcrumbs__ExpandButton-sc-d5msik-3 cOtvQf\\"
            title=\\"Exibir todas as opções\\"
            type=\\"button\\"
          >
            ...
          </button>
        </li>
        <li
-         aria-hidden=\\"true\\"
+         aria-hidden=\\"false\\"
          class=\\"Breadcrumbs__BreadcrumbItem-sc-d5msik-1 cWkJmI\\"
        >
          <svg
            aria-hidden=\\"true\\"
            class=\\"MuiSvgIcon-root Breadcrumbs__BreadcrumbIcon-sc-d5msik-2 jEpugD\\"
@@ -81,11 +82,11 @@
          >
            Quantum
          </a>
        </li>
        <li
-         aria-hidden=\\"true\\"
+         aria-hidden=\\"false\\"
          class=\\"Breadcrumbs__BreadcrumbItem-sc-d5msik-1 cWkJmI\\"
        >
          <svg
            aria-hidden=\\"true\\"
            class=\\"MuiSvgIcon-root Breadcrumbs__BreadcrumbIcon-sc-d5msik-2 jEpugD\\""
`;
